core/objectaction/object.go

```diff
@@ -938,3 +977,74 @@ func (t T) waitExpectation(ctx context.Context, c *client.T, expectation string,
 		}
 	}()
 }
+
+func (t T) waitRequesterSessionEnd(ctx context.Context, c *client.T, requesterSid uuid.UUID, p naming.Path, errC chan<- error) {
+	var (
+		filters []string
+		msg     pubsub.Messager
+
+		err      error
+		evReader event.ReadCloser
+	)
+	filters = []string{
+		fmt.Sprintf("ObjectStatusDeleted,path=%s", p),
+		fmt.Sprintf("ExecFailed,path=%s,requester_sid=%s", p, requesterSid),
+		fmt.Sprintf("ExecSuccess,path=%s,requester_sid=%s", p, requesterSid),
+	}
+	getEvents := c.NewGetEvents().SetFilters(filters)
+	if t.WaitDuration > 0 {
+		getEvents = getEvents.SetDuration(t.WaitDuration)
+	}
+	evReader, err = getEvents.GetReader()
+	if err != nil {
+		return
+	}
+
+	if x, ok := evReader.(event.ContextSetter); ok {
+		x.SetContext(ctx)
+	}
+	go func() {
+		defer func() {
+			if err != nil {
+				err = fmt.Errorf("wait requester session end failed on object %s: %w", p, err)
+			}
+			select {
+			case <-ctx.Done():
+			case errC <- err:
+			}
+		}()
+
+		go func() {
+			// close reader when ctx is done
+			select {
+			case <-ctx.Done():
+				_ = evReader.Close()
+			}
+		}()
+		for {
+			ev, readError := evReader.Read()
+			if readError != nil {
+				if errors.Is(readError, io.EOF) {
+					err = fmt.Errorf("no more events, wait %v failed: %w", p, err)
+				} else {
+					err = readError
+				}
+				return
+			}
+			msg, err = msgbus.EventToMessage(*ev)
+			if err != nil {
+				return
+			}
+			switch m := msg.(type) {
+			case *msgbus.ExecSuccess:
+				return
+			case *msgbus.ExecFailed:
+				errC <- errors.New(m.ErrS)
```

Comment:
possible goroutine leak bug here: should be replaced by err = errors.New(m.ErrS)
defer has a select `case errC <- err:`